drop-transfer/src/ws/server/v6.rs

```diff
@@ -667,12 +673,130 @@ impl handler::Request for (prot::TransferRequest, IpAddr, Arc<DropConfig>) {
     fn parse(self) -> anyhow::Result<IncomingTransfer> {
         let (prot::TransferRequest { files, id }, peer, config) = self;
 
-        let files = files
-            .into_iter()
-            .map(|f| FileToRecv::new(f.id, f.path, f.size))
-            .collect();
-
-        IncomingTransfer::new_with_uuid(peer, files, id, &config)
+        IncomingTransfer::new_with_uuid(peer, map_files(files)?, id, &config)
             .context("Failed to crate transfer")
     }
 }
+
+fn map_files(files: Vec<prot::File>) -> anyhow::Result<Vec<FileToRecv>> {
+    let mut out = Vec::with_capacity(files.len());
+
+    let mut used_mappings = HashMap::new();
+
+    for prot::File { mut path, id, size } in files {
+        let uroot = path.root();
+        let nroot = utils::normalize_filename(uroot);
+
+        'vars: for nvariant in utils::filepath_variants(nroot.as_ref())?
+            .filter_map(|p| p.into_os_string().into_string().ok())
+        {
+            let nroot = match used_mappings.entry(nvariant) {
+                Entry::Occupied(occ) => {
+                    if occ.get() == uroot {
+                        // Good we known the root
+                        occ.key().to_string()
+                    } else {
+                        // The mapping is occupied by other root dir or file.
+                        continue;
+                    }
+                }
+                Entry::Vacant(vacc) => {
+                    // New mapping, lets insert it
+                    let nroot = vacc.key().to_string();
+                    vacc.insert(uroot.to_string());
+                    nroot
+                }
+            };
+
+            let mut piter = path.iter_mut();
+            *piter.next().context("Subpath should always contain root")? = nroot;
+            piter.for_each(|s| *s = utils::normalize_filename(&*s));
+
+            out.push(FileToRecv::new(id, path, size));
+            break 'vars;
+        }
+    }
```

Comment:
I don't really see the point of using `filter_map` and then label with a `break`. Can this be done in more `rust` way or a bit simpler? The logic in this code is quite convoluted due to the problem being such so simplifying it would be beneficial

Reply:
I've removed the label, but I don't see how I can simplify it further. The problem itself is not so straightforward